codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -496,19 +520,49 @@ private void incrementPendingBytes(int numBytes) {
              */
             int write(int allowedBytes) {
                 int before = payload.size();
-                needFlush |= payload.write(Math.max(0, allowedBytes));
-                int writtenBytes = before - payload.size();
+                int writtenBytes = 0;
                 try {
-                    connectionState().incrementStreamWindow(-writtenBytes);
-                    incrementStreamWindow(-writtenBytes);
+                    boolean writeError = true;
+                    try {
+                        writing = true;
```

Comment:
A few things to consider with these changes:
1. This method should not allow re-entrace now. For example if `payload.write(..)` ends up resulting in this method being called again then we will reset `writing` and `canceled` and get confused.  I don't think this is an issue now, but may be good to avoid this complication.
2. `clear` (now `cancel`) was clearing and erroring out the entire queue and now it looks if we are `writing` only the element that is being written will get cleaned up.

What about we push the state into the `FlowControlled` interface (names are just for demonstration :)):
```java
  enum State { IDLE, CANCELED, WRITTING };

  /**
   * Used by the flow controller to track state relative to this payload
   */
  void state(State state);

  /**
   * The flow controller uses this
   */
  State state();
```
The flow controller can then set and check the state before writing and canceling.